proposals/0025/README.md

```diff
@@ -0,0 +1,199 @@
+
+```protobuf
+service WorkflowService {
+  rpc GetWorkflowContexts(WorkflowContextRequest) returns (stream WorkflowContext) {}
+  rpc GetWorkflowActions(WorkflowActionsRequest) returns (WorkflowActionList) {}
+  rpc ReportActionStatus(WorkflowActionStatus) returns (Empty) {}
+  rpc GetWorkflowData(GetWorkflowDataRequest) returns (GetWorkflowDataResponse) {}
+  rpc UpdateWorkflowData(UpdateWorkflowDataRequest) returns (Empty) {}
+}
+```
+
+The structure of the Kubernetes CRD types will start out with those defined in the [Tinkerbell CAPI Provider][capt-types]. The code defining those types would be migrated into the tinkerbell/tink GitHub repository so other clients could import the Go packages.
+
+[capt-types]: https://github.com/tinkerbell/cluster-api-provider-tink/tree/main/tink/api/v1alpha1
+
+## Threat Model
+
+### Actors
+
+**Machines**
+
+In this proposal, a machine is not trusted any more than in Tinkerbell's API today.
+By design, in this proposal machines will have decreased access than they do today, with access to the APIs listed above.
+As it is in Tinkerbell today, the network is trusted when provisioning hardware.
+
+**Hegel**
+
+Hegel will only require read-only access to Hardware CRDs in Kubernetes.
+RBAC policies would govern the level of access to the Kubernetes API.
+
+Hegel will need to guard against confused-deputy attacks and not return metadata to the wrong client.
+
+**Boots**
+
+Boots will only require read-only access to Hardware CRDs in Kubernetes.
+RBAC policies would govern the level of access to the Kubernetes API.
+
+**PBnJ**
+
+PBnJ will require:
+* Read-only access to Hardware CRDs to discover management interface connectivity
+* Mutating access to Workflow CRDs to execute workflow steps like power cycling and BIOS management
+* Some level of secret access to connect to management interfaces. Implementation and design of that access is out of scope of this proposal.
+
+**Tink API**
+
+The Tinkerbell API will require:
+* Read access to Hardware CRDs in Kubernetes
+* Read access to Template CRDs in Kubernetes
+* Write access to Workflow CRDs in Kubernetes
+
+**Tinkerbell Workflow Controller**
+
+There will need to be a controller to process creation of workflows. It will need:
+* Read access to Hardware CRDs in Kubernetes
+* Read access to Template CRDs in Kubernetes
+* Write access to Workflow CRDs `.spec` in Kubernetes
+
+**Tinkerbell Validating Webhook**
+
+A validating webhook will be used to validate Templates and Workflows as they are created.
+It will not require any Kubernetes API access, but needs to be reachable by the Kubernetes API server.
+
+**Tinkerbell Administrator**
+
+Tinkerbell administrators might require varying levels of access in order to perform CRUD operations on any or all Tinkerbell types.
+Kubernetes cluster administrators can define custom levels of access with RBAC policies to grant Tinkerbell admins the correct least-privilege level of access.
+
+## Alternatives
```

Comment:
One thing to consider in this doc is how to safely transition existing users of Tinkerbell to this model. We can't suddenly cut-over to the new API and break existing users. Similar to Kubernetes itself, we probably need to hide this new feature work behind a flag for some time: particularly if the previous behavior is to be deprecated.

Reply:
I added a migration section